flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,325 @@
+      // argument.
+      field.future = future;
+      field.dep = flecsi::data::dependency::write;
+    }
+    else if(privilege_read(P)) {
+      // request future from promise only if required (once for all arguments)
+      if(!future.valid()) {
+        future = promise.get_future();
+      }
+
+      // If the task reads from the current argument then we must associate
+      // the future that represents the end of the task execution with the
+      // current argument. We have to make sure that possibly more than one
+      // read operation may have to finish before other operations are allowed
+      // to go ahead.
+      if(!field.future.valid() || field.future.is_ready()) {
+        // This is either the first operation using the given field or any
+        // previous operations have already finished.
+        field.future = future;
+      }
+      else if(field.dep != flecsi::data::dependency::write) {
+        // This read operation needs to be added to the list of (read-)
+        // dependencies already existing for the given field.
+        field.future = hpx::when_all(std::move(field.future), future).share();
+      }
+      field.dep = flecsi::data::dependency::read;
+    }
+  }
+
+  // visit for reduction operation
+  template<typename R,
+    typename T,
+    typename Topo,
+    typename Topo::index_space Space>
+  void visit(data::reduction_accessor<R, T> &,
+    const data::field_reference<T, data::dense, Topo, Space> & ref) {
+    static_assert(std::is_same_v<typename Topo::base, topo::global_base>);
+    const field_id_t f = ref.fid();
+    auto & r = *ref.topology();
+
+    // store associated region for bind_accessors
+    regions.push_back(&r);
+    partitions.push_back(nullptr);
+
+    // Reduction input must be treated as a "read/write after write" dependency
+    auto & field = r[f];
+
+    if(field.future.valid()) {
+      flog_assert(field.dep == data::dependency::write,
+        "field reference must represent a 'read after write' dependency");
+      dependencies.push_back(field.future);
+    }
+
+    // request future from promise only if required (once for all arguments)
+    if(!future.valid()) {
+      future = promise.get_future();
+    }
+
+    // Reductions write to the current argument, thus we must associate the
+    // future that represents the end of the task execution with the current
+    // argument.
+    field.future = future;
+    field.dep = flecsi::data::dependency::write;
+  }
+
+public:
+  // Delay the execution of the given task until all dependencies have been
+  // satisfied (if any).
+  template<typename R, typename Params, typename Task>
+  ::hpx::future<R>
+  delay_execution(Params && params, std::string task_name, Task && task) && {
+
+    ::hpx::future<R> result;
+    if(dependencies.empty()) {
+      auto finalize = param_buffers(params, task_name);
+
+      // directly execute task as no dependencies have to be taken into account
```

Comment:
Maybe it doesn't matter because ghost copies are always deferred, but if there weren't any, would we just execute all tasks directly because each would have necessarily had its dependencies satisfied by all the direct executions before it?

Reply:
That's the idea, yes.

Reply:
What did that mean before ghost copies were made asynchronous here?  Wouldn't _everything_ run synchronously?